### PR TITLE
Better request handling

### LINKED ISSUE #8231: [BUG] [1.16.5] Issues with the Cook/Restaurant
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest MineColonies Verison?

- [X] I am running the latest **alpha** version of MineColonies for my Minecraft version.

### Did you check on the Wiki? or ask on Discord?

- [X] I checked the MineColonies Wiki and made sure my issue is not covered there. Or I was sent from discord to open an issue here.

### Minecraft Version

1.16

### MineColonies Version

1.16.5-1.0.580-ALPHA

### Structurize Version

1.16.5-1.0.371-ALPHA

### Related Mods and their Versions

https://gist.github.com/Sandriell/0688b4e8e61c46e1bd44c36ad9ff6fc0

### Current Behavior

Two issues with the Cook/Assistant Cook...

1. Similar to the [previously reported issue](https://github.com/ldtteam/minecolonies/issues/8103) with the Assistant Cook keeping what he makes, the cook is doing the same thing- resulting in endless loops with minimum stock requirements not being met. I watched him make several stacks before interfering and moving one of them to the hut's rack and stopping the loop. Though thinking about this maybe the real issue is that the cook is making the food in the first place, and the Assistant Cook should be doing it? These are taught recipes as well, not simple cooked food.
2.  They fail to keep track of ingredients they already have. I am not sure exactly where in the process this is breaking down, but the inventory for the Restaurant will fill up with ingredients over time that the cook completely ignores, thus requesting even more. This is the issue that completely breaks food from mods like Pam's Harvestcraft because they fail to see the utensils they already have.

### Expected Behavior

Cook puts what he makes in the inventory, or what is in his inventory counts towards the minimum stock since he is the one handing out the food anyway.

Using the ingredients that are already in the hut inventory first before requesting more.

### Reproduction Steps

1. Teach the restaurant a recipe.
2. Set a minimum stock for that item.
3. Wait for/supply all the ingredients - sometimes my manually moving them back to the warehouse so they are "seen" again.
4. Cook cook makes the request
5. Immediately requests to make another stock to try and satisfy the minimum stock requirement

### Logs

N/A

### Anything else?

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #8245: [BUG] Courrier not delivering ressources to builder / not doing pickups
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest MineColonies Verison?

- [X] I am running the latest **alpha** version of MineColonies for my Minecraft version.

### Did you check on the Wiki? or ask on Discord?

- [X] I checked the MineColonies Wiki and made sure my issue is not covered there. Or I was sent from discord to open an issue here.

### Minecraft Version

1.18

### MineColonies Version

minecolonies-1.18.2-1.0.585-BETA

### Structurize Version

structurize-1.18.2-1.0.370-ALPHA

### Related Mods and their Versions

Forge 40.0.44
blockui-1.18.2-0.0.47-ALPHA
domum_ornamentum-1.18.2-1.0.48-ALPHA-universal
multi-piston-1.18.2-1.2.3-ALPHA

### Current Behavior

Courriers are unable to deliver results of craft to builders and are unable to perform pickups of several huts (blacksmith, stonesmelter and stonemason). They are able to deliver resources from the wharehouse to the huts but not to pickup the finished goods. They are able to do pickups of other huts (it works for builders) but not of the huts listed above.

image

### Expected Behavior

Courrier should be able to deliver the resources and to perform the pickups

### Reproduction Steps

1. Builder request resources
2. Stonemason / Stonesmelter / Blacksmith perform craft
3. Courrier doesn't deliver resources from those huts

1. Request pickup of a hut
2. No message appears in the chat
3. Request pickup again
4. "unable to comply, a pickup is already in progress"
5. No courrier comes for a pickup

### Logs

https://pastebin.com/gH2eb8zs

### Anything else?

Paste is latest.log of the 1.18.1 version (I play on the modpack) but I tested the issue in the lastest version of minecolonies and it happened as well

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #8245
Closes #8231
Closes #

# Changes proposed in this pull request:
- This PR introduces resolver suitability. Resolvers can return a metric of how suited they are for the task.
- This could be distance-based but also based on how long the list is.
- It goes through the list of resolvers and for all resovlers of the same type it compares suitability.
- Removes private crafting resolver from cook, only assistant can craft

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/management/handlers/RequestHandler.java [hunk 1]
```diff
@@ -159,54 +162,92 @@ public IToken<?> assignRequestDefault(final IRequest<?> request, final Collectio
...
+                {
+                    previousResolver = resolver;
+                    previousMetric = resolver.getSuitabilityMetric(request);
+                }
                 continue;
             }
 
-            //Successfully found a resolver. Registering
-            manager.getLogger().debug("Finished resolver assignment search for request: " + request + " successfully");
+            if (previousResolver.getClass().equals(resolver.getClass()))
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/management/handlers/RequestHandler.java:177]: This logic seems weird to me.
Why do you restrict this logic to the same class?

**Comment by marchermans** [comment]
The implementation here is a nice idea in concept but the implementation feels weird for solving the noted issue, it would have sufficed to not register the resolvers for example when there are no working citizens in said building, or check if there is an assigned delivery man.

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/management/handlers/RequestHandler.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/management/handlers/RequestHandler.java:177]: Fair, I'll adjust this logic to resolvers with the same priority not of the same class

**Comment by Raycoms** [comment]
> The implementation here is a nice idea in concept but the implementation feels weird for solving the noted issue, it would have sufficed to not register the resolvers for example when there are no working citizens in said building, or check if there is an assigned delivery man.

Because it does more. It allows the closest warehouse, the closest crafter to fulfill, or even, the worker with the smallest queue, etc. (Current implementation is solely distance based yet)

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/management/handlers/RequestHandler.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/management/handlers/RequestHandler.java:177]: rethinking this, I think i want to keep it the same class, to guarantee the metric is the same

**Review by marchermans — DISMISSED** [review]

**Review by marchermans — APPROVED** [review]

**Pull request merged**